Домашнее задание 1 - 10 баллов

- Загрузите набор данных lenta-ru-news с помощью библиотеки Corus для задачи классификации текстов по топикам (пригодятся атрибуты title, text, topic)- **1 балл**

- Подготовьте данные к обучению: - **3 балла**
- Предобработайте данные: реализуйте оптимальную, на ваш взгляд, предобработку текстов (нормализация, очистка, стемминг/лемматизация и т.п.) и таргета.
<br> **hint**: для ускорения обработки и обучения можно ограничиться не всем датасетом, а его репрезентативной частью, например, размера 100_000.
- Кратко опишите пайплайн, на котором остановились, и почему.
- Разделите датасет на обучающую, валидационную и тестовую выборки со стратификацией в пропорции 60/20/20. В качестве целевой переменной используйте атрибут topic
- Замерьте базовое качество с любым dummy-бейзлайном - **0.5 балла**
- Обучите модель sklearn.linear_model.LogisticRegression с двумя вариантами векторизации: **2 балла**
<br> sklearn.feature_extraction.text.CountVectorizer <br>
sklearn.feature_extraction.text.TfidfVectorizer
- Попробуйте улучшить качество, подобрав оптимальные гиперпараметры трансформаций и модели на кросс-валидации **1 балл**
- Оцените качество лучшего пайплайна на отложенной выборке - **0.5 балла**
<br>

**Общее**
- Принимаемые решения обоснованы (почему выбрана определенная архитектура/гиперпараметр/оптимизатор/преобразование и т.п.) - **1 балл**
- Обеспечена воспроизводимость решения: зафиксированы random_state, ноутбук воспроизводится от начала до конца без ошибок - **1 балл**

In [ ]:
%%capture
!pip install corus

In [ ]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz

--2025-03-01 19:12:51--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz
Resolving github.com (github.com)... 140.82.112.4
Connecting to github.com (github.com)|140.82.112.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250301%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250301T191252Z&X-Amz-Expires=300&X-Amz-Signature=41b48152dca63230e6e5c947618895a1078cec9c0b9cb930805de207534b2b47&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.gz&response-content-type=application%2Foctet-stream [following]
--2025-03-01 19:12:52--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-

In [ ]:
%%capture
!pip install pymorphy3

In [ ]:
%%capture
!pip install clean-text

In [ ]:
import pandas as pd
import numpy as np
from corus import load_lenta
import random

from cleantext import clean
from pymorphy3 import MorphAnalyzer
from bs4 import BeautifulSoup
import re
from nltk.corpus import stopwords
import nltk
nltk.download('stopwords')

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import LabelEncoder

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
path = 'lenta-ru-news.csv.gz'
records = load_lenta(path)
next(records)

LentaRecord(
    url='https://lenta.ru/news/2018/12/14/cancer/',
    title='Названы регионы России с\xa0самой высокой смертностью от\xa0рака',
    text='Вице-премьер по социальным вопросам Татьяна Голикова рассказала, в каких регионах России зафиксирована наиболее высокая смертность от рака, сообщает РИА Новости. По словам Голиковой, чаще всего онкологические заболевания становились причиной смерти в Псковской, Тверской, Тульской и Орловской областях, а также в Севастополе. Вице-премьер напомнила, что главные факторы смертности в России — рак и болезни системы кровообращения. В начале года стало известно, что смертность от онкологических заболеваний среди россиян снизилась впервые за три года. По данным Росстата, в 2017 году от рака умерли 289 тысяч человек. Это на 3,5 процента меньше, чем годом ранее.',
    topic='Россия',
    tags='Общество',
    date=None
)

In [ ]:
df_sample=pd.read_csv('sample_lenta.csv')

### Обеспечение воспроизводимости ноутбука (random_state)

In [ ]:
RANDOM_STATE = 42

def set_random_seed():
    np.random.seed(RANDOM_STATE)
    random.seed(RANDOM_STATE)

set_random_seed()

### Выбор 100000 случайных оъектов, преобразование формата в датафрейм

Ограничиваем выборку до 100 000 записей, чтобы ускорить обработку и обучение модели. Датасет lenta-ru-news очень большой, и работать со всей выборкой слишком затратно по ресурсам.



In [ ]:
data = []
for record in records:
    data.append({
        'title': record.title,
        'text': record.text,
        'topic': record.topic
    })

df = pd.DataFrame(data)

df_sample = df.sample(n=100000, random_state=RANDOM_STATE)
df_sample.to_csv('sample_lenta.csv')

### Предобработка текста

Тексты могут содержать много мусора, который не несёт смысла для модели, но увеличивает размерность данных и снижает качество работы модели. Библиотека clean-text была выбрана, потому что она автоматизирует множество стандартных задач предобработки текста.

Лемантизация выбрана для обработки слов, так как русский язык довольно сложный по структуре и лемантизация показывает лучшие результаты нежели стемминг. Для английского языка используется чаще второй вариант. pymorphy2 достаточно точно анализирует сложные слова, поддерживает временные формы и склонения, что делает его более надёжным в реальной задаче.

Также удалим слишком короткие слова и стоп-слова, так как они часто не несут смысловой нагрузки.

In [ ]:
morph = MorphAnalyzer()
stop_words = set(stopwords.words('russian'))

def preprocess_text(text):
    if pd.isna(text):
        return ''

    # Чистка текста с clean-text
    text = clean(text,
                 fix_unicode=True,
                 to_ascii=False,
                 lower=True,
                 no_line_breaks=True,
                 no_urls=True,
                 no_emails=True,
                 no_phone_numbers=True,
                 no_numbers=True,
                 no_digits=True,
                 no_currency_symbols=True,
                 no_punct=True,
                 replace_with_url="<URL>",
                 replace_with_email="<EMAIL>",
                 replace_with_phone_number="<PHONE>",
                 replace_with_number="<NUMBER>",
                 replace_with_digit="0",
                 replace_with_currency_symbol="<CUR>",
                 lang="ru")

    # Удаление лишних пробелов
    text = re.sub(r'\s+', ' ', text).strip()

    # Лемматизация и удаление стоп-слов
    words = text.split()
    words = [morph.parse(word)[0].normal_form for word in words if word not in stop_words and len(word) > 2]

    return ' '.join(words)

df_sample['text'] = df_sample['text'].apply(preprocess_text)

df_sample = df_sample[df_sample['text'].str.strip() != '']

### Кодирование целевой переменной

topic - это строковый признак, а модель требует числовые значения.
Мы используем LabelEncoder, чтобы преобразовать категории в числа.

In [ ]:
label_encoder = LabelEncoder()
df_sample['topic'] = label_encoder.fit_transform(df_sample['topic'])

### Разделение на train/val/test

Train (60%) - обучение модели.
<br>Validation (20%) - подбор гиперпараметров.
<br>Test (20%) - финальная проверка.

In [ ]:
train_data, test_data = train_test_split(df_sample, test_size=0.4, random_state=42)
val_data, test_data = train_test_split(test_data, test_size=0.5, random_state=42)

### Базовый dummy-классификатор

Dummy-классификатор просто предсказывает самый частый класс.

In [ ]:
dummy = DummyClassifier(strategy='most_frequent')
dummy.fit(train_data['text'], train_data['topic'])
dummy_pred = dummy.predict(test_data['text'])
print(f"Dummy accuracy: {accuracy_score(test_data['topic'], dummy_pred):.4f}")

Dummy accuracy: 0.2161


### Векторизация и обучение логистической регрессии

In [ ]:
vectorizers = {'CountVectorizer': CountVectorizer(ngram_range=(1,2), max_features=50000),
               'TfidfVectorizer': TfidfVectorizer(ngram_range=(1,2), max_features=50000)}

for name, vectorizer in vectorizers.items():
    pipeline = Pipeline([
        ('vectorizer', vectorizer),
        ('classifier', LogisticRegression(max_iter=1000, random_state=42))
    ])

    pipeline.fit(train_data['text'], train_data['topic'])
    preds = pipeline.predict(test_data['text'])
    acc = accuracy_score(test_data['topic'], preds)
    print(f"{name} accuracy: {acc:.4f}")

CountVectorizer accuracy: 0.8031
TfidfVectorizer accuracy: 0.8072


Без подбора гиперпараметров TfIdfVectorizer показывает лучшие результаты.

### Подбор оптимальных гиперпараметров

max_features: Ограничиваем число фичей (20000 и 50000). Если словарь слишком большой, модель может переобучиться.
<br>C: Гиперпараметр регуляризации LogisticRegression. Чем больше C, тем меньше регуляризация (то есть модель может переобучиться).
<br>solver:
liblinear - работает лучше для небольших датасетов.
lbfgs - лучше справляется с большим количеством классов.


In [ ]:
param_grid = {
    'vectorizer__ngram_range': [(1,2)],          # Униграммы и биграммы
    'vectorizer__max_features': [20000, 50000],  # Размер словаря
    'classifier__C': [0.1, 1],                   # Регуляризация логистической регрессии
    'classifier__solver': ['liblinear', 'lbfgs'] # Разные методы оптимизации
}

results = {}

for name, vectorizer in vectorizers.items():
    print(f'Оптимизация для {name}')

    pipeline = Pipeline([
        ('vectorizer', vectorizer),
        ('classifier', LogisticRegression(max_iter=1000, random_state=42))
    ])

    gs = GridSearchCV(pipeline, param_grid, cv=3, n_jobs=-1, scoring='accuracy', verbose=1)
    gs.fit(train_data['text'], train_data['topic'])

    best_model = gs.best_estimator_
    preds = best_model.predict(test_data['text'])

    acc = accuracy_score(test_data['topic'], preds)
    results[name] = {
        'best_params': gs.best_params_,
        'accuracy': acc,
        'classification_report': classification_report(test_data['topic'], preds)
    }

for name, res in results.items():
    print(f'{name} - Лучшая модель:')
    print(f'Best params: {res["best_params"]}')
    print(f'Accuracy: {res["accuracy"]:.4f}')
    print('Classification Report:')
    print(res['classification_report'])
    print('')


Оптимизация для CountVectorizer
Fitting 3 folds for each of 8 candidates, totalling 24 fits


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=3.
  warnings.warn(


Оптимизация для TfidfVectorizer
Fitting 3 folds for each of 8 candidates, totalling 24 fits


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/m

CountVectorizer - Лучшая модель:
Best params: {'classifier__C': 0.1, 'classifier__solver': 'liblinear', 'vectorizer__max_features': 50000, 'vectorizer__ngram_range': (1, 2)}
Accuracy: 0.8155
Classification Report:
              precision    recall  f1-score   support

           0       0.73      0.23      0.35        35
           1       0.00      0.00      0.00         2
           2       0.67      0.34      0.45       201
           3       0.83      0.82      0.83      1420
           4       0.87      0.78      0.82       611
           5       0.67      0.60      0.64       696
           6       0.78      0.72      0.75      1242
           7       0.67      0.33      0.44        18
           8       0.00      0.00      0.00        12
           9       0.87      0.88      0.88      1494
          10       1.00      1.00      1.00         2
          12       0.80      0.84      0.82      3735
          13       0.84      0.83      0.84      1440
          14       0.80      

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Лучшую точность показал CountVectorizer: 0.8155.

### Оценка на тесте для лучших моделей для каждого векторайзера

In [ ]:
for name, res in results.items():
    best_model = res['best_params']
    print(f'Запуск на тестовой выборке для {name}')

    final_pipeline = Pipeline([
        ('vectorizer', vectorizers[name].set_params(**{
            'ngram_range': best_model['vectorizer__ngram_range'],
            'max_features': best_model['vectorizer__max_features']
        })),
        ('classifier', LogisticRegression(
            C=best_model['classifier__C'],
            solver=best_model['classifier__solver'],
            max_iter=1000,
            random_state=42
        ))
    ])

    final_pipeline.fit(train_data['text'], train_data['topic'])
    test_preds = final_pipeline.predict(test_data['text'])

    acc = accuracy_score(test_data['topic'], test_preds)
    print(f'Итоговая Accuracy: {acc:.4f}')
    print('Classification Report:')
    print(classification_report(test_data['topic'], test_preds))
    print('')

Запуск на тестовой выборке для CountVectorizer
Итоговая Accuracy: 0.8155
Classification Report:
              precision    recall  f1-score   support

           0       0.73      0.23      0.35        35
           1       0.00      0.00      0.00         2
           2       0.67      0.34      0.45       201
           3       0.83      0.82      0.83      1420
           4       0.87      0.78      0.82       611
           5       0.67      0.60      0.64       696
           6       0.78      0.72      0.75      1242
           7       0.67      0.33      0.44        18
           8       0.00      0.00      0.00        12
           9       0.87      0.88      0.88      1494
          10       1.00      1.00      1.00         2
          12       0.80      0.84      0.82      3735
          13       0.84      0.83      0.84      1440
          14       0.80      0.60      0.69       165
          15       0.78      0.84      0.81      4322
          16       0.70      0.48      

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Итоговая Accuracy: 0.8072
Classification Report:
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        35
           1       0.00      0.00      0.00         2
           2       0.76      0.13      0.22       201
           3       0.82      0.80      0.81      1420
           4       0.87      0.74      0.80       611
           5       0.70      0.56      0.62       696
           6       0.78      0.69      0.73      1242
           7       1.00      0.11      0.20        18
           8       0.00      0.00      0.00        12
           9       0.86      0.88      0.87      1494
          10       0.00      0.00      0.00         2
          12       0.79      0.85      0.82      3735
          13       0.82      0.84      0.83      1440
          14       0.80      0.52      0.63       165
          15       0.76      0.85      0.80      4322
          16       0.73      0.32      0.45       540
          17       0.95      0.9

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


На тестовой выборке получилась такая же точность:
<br>**CountVectorizer: 0.8155**
<br>TfIdfVectorizer: 0.8072